pkg/cmd/server/kubernetes/master/controller.go

```diff
@@ -0,0 +1,311 @@
+	if err != nil {
+		return false, fmt.Errorf("A backward-compatible provisioner could not be created: %v, but one was expected. Provisioning will not work. This functionality is considered an early Alpha version.", err)
+	}
+
+	eventcast := record.NewBroadcaster()
+	recorder := eventcast.NewRecorder(kapi.Scheme, kclientv1.EventSource{Component: bootstrappolicy.InfraPersistentVolumeControllerServiceAccountName})
+	eventcast.StartRecordingToSink(
+		&kv1core.EventSinkImpl{
+			Interface: ctx.ClientBuilder.ClientGoClientOrDie(bootstrappolicy.InfraPersistentVolumeControllerServiceAccountName).CoreV1().Events(""),
+		},
+	)
+	recyclerSA := bootstrappolicy.InfraPersistentVolumeRecyclerControllerServiceAccountName
+	plugins, err := probeRecyclableVolumePlugins(ctx.Options.VolumeConfiguration, c.OpenShiftInfrastructureNamespace, c.RecyclerImage, recyclerSA)
+	if err != nil {
+		return false, err
+	}
+
+	volumeController := persistentvolumecontroller.NewController(
+		persistentvolumecontroller.ControllerParameters{
+			KubeClient:                ctx.ClientBuilder.ClientOrDie(bootstrappolicy.InfraPersistentVolumeControllerServiceAccountName),
+			SyncPeriod:                ctx.Options.PVClaimBinderSyncPeriod.Duration,
+			AlphaProvisioner:          alphaProvisioner,
+			VolumePlugins:             plugins,
+			Cloud:                     c.CloudProvider,
+			ClusterName:               ctx.Options.ClusterName,
+			VolumeInformer:            ctx.InformerFactory.Core().V1().PersistentVolumes(),
+			ClaimInformer:             ctx.InformerFactory.Core().V1().PersistentVolumeClaims(),
+			ClassInformer:             ctx.InformerFactory.Storage().V1beta1().StorageClasses(),
+			EventRecorder:             recorder,
+			EnableDynamicProvisioning: ctx.Options.VolumeConfiguration.EnableDynamicProvisioning,
+		})
+
+	go volumeController.Run(ctx.Stop)
+
+	return true, nil
+}
+
+type PersistentVolumeAttachDetachControllerConfig struct {
+	CloudProvider cloudprovider.Interface
+}
+
+func (c *PersistentVolumeAttachDetachControllerConfig) RunController(ctx kubecontroller.ControllerContext) (bool, error) {
+	attachDetachController, err := attachdetachcontroller.NewAttachDetachController(
+		ctx.ClientBuilder.ClientOrDie(bootstrappolicy.InfraPersistentVolumeAttachDetachControllerServiceAccountName),
+		ctx.InformerFactory.Core().V1().Pods(),
+		ctx.InformerFactory.Core().V1().Nodes(),
+		ctx.InformerFactory.Core().V1().PersistentVolumeClaims(),
+		ctx.InformerFactory.Core().V1().PersistentVolumes(),
+		c.CloudProvider,
+		kctrlmgr.ProbeAttachableVolumePlugins(ctx.Options.VolumeConfiguration),
+		ctx.Options.DisableAttachDetachReconcilerSync,
+		ctx.Options.ReconcilerSyncLoopPeriod.Duration,
+	)
+	if err != nil {
+		return false, fmt.Errorf("failed to start attach/detach persistent volume controller: %v", err)
+	}
+
+	go attachDetachController.Run(ctx.Stop)
+
+	return true, nil
+}
+
+type SchedulerControllerConfig struct {
+	PrivilegedClient               kclientset.Interface
```

Comment:
Let's get a TODO on this and an issue.  I'd like to move closer to upstream.

Do we have a `kubeorigin` label yet?  I think we need to start tracking all the various refactors we need.